mllib/src/test/scala/org/apache/spark/ml/tuning/RandomRangesSuite.scala

```diff
@@ -0,0 +1,167 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.ml.tuning
+
+import scala.reflect.runtime.universe.TypeTag
+
+import org.scalacheck.{Arbitrary, Gen}
+import org.scalacheck.Arbitrary._
+import org.scalacheck.Gen.Choose
+import org.scalatest.{Assertion, Succeeded}
+import org.scalatest.matchers.must.Matchers
+import org.scalatestplus.scalacheck.ScalaCheckDrivenPropertyChecks
+
+import org.apache.spark.SparkFunSuite
+
+class RandomRangesSuite extends SparkFunSuite with ScalaCheckDrivenPropertyChecks with Matchers {
+
+  import RandomRanges._
+
+  test("log of any base") {
+    assert(logN(16, 4) == 2d)
+    assert(logN(1000, 10) === (3d +- 0.000001))
+    assert(logN(256, 2) == 8d)
+  }
+
+  test("random doubles in log space") {
+    val gen: Gen[(Double, Double, Int)] = for {
+      x <- Gen.choose(0d, Double.MaxValue)
+      y <- Gen.choose(0d, Double.MaxValue)
+      n <- Gen.choose(0, Int.MaxValue)
+    } yield (x, y, n)
+    forAll(gen) { case (x, y, n) =>
+      val lower = math.min(x, y)
+      val upper = math.max(x, y)
+      val result = randomLog(x, y, n)
+      assert(result >= lower && result <= upper)
+    }
+  }
+
+  test("random BigInt generation does not go into infinite loop") {
+    assert(randomBigInt0To(0) == BigInt(0))
+  }
+
+  test("random ints") {
+    checkRange(Linear[Int])
+  }
+
+  test("random log ints") {
+    checkRange(Log10[Int])
+  }
+
+  test("random int distribution") {
+    checkDistributionOf(1000)
+  }
+
+  test("random doubles") {
+    checkRange(Linear[Double])
+  }
+
+  test("random log doubles") {
+    checkRange(Log10[Double])
+  }
+
+  test("random double distribution") {
+    checkDistributionOf(1000d)
+  }
+
+  test("random floats") {
+    checkRange(Linear[Float])
+  }
+
+  test("random log floats") {
+    checkRange(Log10[Float])
+  }
+
+  test("random float distribution") {
+    checkDistributionOf(1000f)
+  }
+
+  abstract class RandomFn[T: Numeric: Generator] {
```

Comment:
These defs below can probably be private; it's not a big deal

Reply:
Done.